File: go/pkg/speech/segment.go
package speech

import (
	"bytes"
	"io"
	"slices"
	"sync"
	"unicode"
	"unicode/utf8"

	"google.golang.org/api/iterator"
)

var (
	_ SentenceSegmenter = (*DefaultSentenceSegmenter)(nil)
	_ SentenceIterator  = (*defaultSentenceIterator)(nil)
)

// DefaultSentenceSegmenter is the default sentence segmenter.
type DefaultSentenceSegmenter struct {
	// MaxRunesPerSegment specifies the maximum number of runes allowed in each
	// segment. If the text exceeds this limit, it will be split into multiple
	// segments. If this value is not set, it defaults to 256.
	MaxRunesPerSegment int
}

// Segment segments the given reader into sentences.
func (s DefaultSentenceSegmenter) Segment(r io.Reader) (SentenceIterator, error) {
	iter := &defaultSentenceIterator{
		writeNotify: make(chan struct{}, 1),
		buf:         bytes.NewBuffer(nil),
	}
	if s.MaxRunesPerSegment > 0 {
		iter.maxRunesPerSegment = s.MaxRunesPerSegment
	} else {
		iter.maxRunesPerSegment = 256
	}

	go func() {
		defer close(iter.writeNotify)

		// Copy to iter not buf for thread safety
		if _, err := io.Copy(iter, r); err != nil {
			iter.closeWithError(err)
		}
	}()
	return iter, nil
}

type defaultSentenceIterator struct {
	maxRunesPerSegment int

	mu               sync.Mutex
	closed           bool
	firstSegReturned bool
	err              error
	writeNotify      chan struct{}
	buf              *bytes.Buffer
}

func (s *defaultSentenceIterator) Write(p []byte) (int, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closed {
		return 0, io.ErrClosedPipe
	}
	n, err := s.buf.Write(p)
	if err != nil {
		return n, err
	}
	select {
	case s.writeNotify <- struct{}{}:
	default:
	}
	return n, nil
}

func (s *defaultSentenceIterator) nextRunes(move bool) (b []byte, full bool) {
	if move {
		defer func() {
			s.buf.Next(len(b))
		}()
	}
	b = s.buf.Bytes()
	idx := lastRuneIndex(b)
	b = b[:idx]
	if rs := []rune(string(b)); len(rs) >= s.maxRunesPerSegment {
		b = []byte(string(rs[:s.maxRunesPerSegment]))
		full = true
	}
	return
}

func (s *defaultSentenceIterator) Next() (string, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	defer func() {
		s.firstSegReturned = true
	}()

	eof := false

	for {
		if s.closed {
			if s.err != nil {
				return "", s.err
			}
			return "", iterator.Done
		}
		if eof {
			if s.buf.Len() > 0 {
				b, _ := s.nextRunes(true)
				return string(b), nil
			}
			return "", iterator.Done
		}
		if b, full := s.nextRunes(false); len(b) > 0 {
			var idx int
			if s.firstSegReturned {
				idx = lastSegmentBoundaryIndex(b)
			} else {
				idx = segmentBoundaryIndex(b)
			}
			switch {
			case idx > 0:
				s.buf.Next(idx)
				return string(b[:idx]), nil
			case idx == 0 && full:
				s.buf.Next(len(b))
				return string(b), nil
			}
		}
		s.mu.Unlock()
		_, ok := <-s.writeNotify
		eof = !ok
		s.mu.Lock()
	}
}

func (s *defaultSentenceIterator) Close() {
	s.closeWithError(nil)
}

func (s *defaultSentenceIterator) closeWithError(err error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.closed = true
	s.err = err
	s.buf.Reset()
}

// lastRuneIndex returns the index of the last rune in the given byte slice.
func lastRuneIndex(b []byte) int {
	for i := len(b) - 1; i >= 0; i-- {
		if b[i] < utf8.RuneSelf {
			return i + 1
		}
		if !utf8.RuneStart(b[i]) {
			continue
		}
		r, sz := utf8.DecodeRune(b[i:])
		if r != utf8.RuneError {
			return i + sz
		}
		return i
	}
	return 0
}

// lastSegmentBoundaryIndex returns the index of the last boundary rune in the
// given byte slice.
func lastSegmentBoundaryIndex(b []byte) int {
	it := lastRuneIndex(b)
	rs := []rune(string(b[:it]))

	n := 0

	for i, r := range slices.Backward(rs) {
		prev := '0'
		if i > 0 {
			prev = rs[i-1]
		}
		next := '0'
		if i < len(rs)-1 {
			next = rs[i+1]
			n += utf8.RuneLen(next)
		}
		switch r {
		case '.', ':', ',', '：':
			// This case handles 9.9 and 10:15
			if unicode.IsNumber(next) && unicode.IsNumber(prev) {
				continue
			}
			fallthrough
		case '，', '；', '。', '？', '！', '…', '～',
			'?', '!', '¿', '¡', ';', '~',
			'\r', '\n', '„', '・':
			return it - n
		}
	}
	return 0
}

// segmentBoundaryIndex returns the index of the first boundary rune in the
// given byte slice.
func segmentBoundaryIndex(b []byte) int {
	it := lastRuneIndex(b)
	rs := []rune(string(b[:it]))

	n := 0
	for i, r := range rs {
		n += utf8.RuneLen(r)
		prev := '0'
		if i > 0 {
			prev = rs[i-1]
		}
		next := '0'
		if i < len(rs)-1 {
			next = rs[i+1]
		}
		switch r {
		case '.', ':', ',', '：':
			// This case handles 9.9 and 10:15
			if unicode.IsNumber(next) && unicode.IsNumber(prev) {
				continue
			}
			fallthrough
		case '，', '；', '。', '？', '！', '…', '～',
			'?', '!', '¿', '¡', ';', '~',
			'\r', '\n', '„', '・':
			return n
		}
	}
	return 0
}

File: rust/speech/src/segment.rs
//! Sentence segmentation utilities.

use std::sync::Arc;
use tokio::io::{AsyncRead, AsyncReadExt};
use tokio::sync::{mpsc, Mutex};

/// Error type for segmentation operations.
#[derive(Debug, thiserror::Error)]
pub enum SegmentError {
    #[error("end of stream")]
    Done,
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
    #[error("closed")]
    Closed,
}

/// Interface for iterating over sentences.
#[async_trait::async_trait]
pub trait SentenceIterator: Send + Sync {
    /// Returns the next sentence.
    async fn next(&mut self) -> Result<String, SegmentError>;

    /// Closes the iterator.
    fn close(&mut self);
}

/// Interface for segmenting text into sentences.
#[async_trait::async_trait]
pub trait SentenceSegmenter: Send + Sync {
    /// Segments the text from the reader into sentences.
    async fn segment(
        &self,
        reader: Box<dyn AsyncRead + Send + Unpin>,
    ) -> Result<Box<dyn SentenceIterator>, SegmentError>;
}

/// Default sentence segmenter that splits on punctuation.
pub struct DefaultSentenceSegmenter {
    /// Maximum number of characters allowed in each segment.
    /// If the text exceeds this limit, it will be split into multiple segments.
    /// Defaults to 256.
    pub max_chars_per_segment: usize,
}

impl Default for DefaultSentenceSegmenter {
    fn default() -> Self {
        Self {
            max_chars_per_segment: 256,
        }
    }
}

impl DefaultSentenceSegmenter {
    /// Creates a new segmenter with the specified max chars per segment.
    pub fn new(max_chars: usize) -> Self {
        Self {
            max_chars_per_segment: max_chars,
        }
    }
}

#[async_trait::async_trait]
impl SentenceSegmenter for DefaultSentenceSegmenter {
    async fn segment(
        &self,
        reader: Box<dyn AsyncRead + Send + Unpin>,
    ) -> Result<Box<dyn SentenceIterator>, SegmentError> {
        let (tx, rx) = mpsc::channel(16);
        let max_chars = self.max_chars_per_segment;

        tokio::spawn(async move {
            let mut reader = reader;
            let mut buf = String::new();
            let mut temp = [0u8; 1024];

            loop {
                match reader.read(&mut temp).await {
                    Ok(0) => {
                        // EOF - send remaining buffer
                        if !buf.is_empty() {
                            let _ = tx.send(Ok(std::mem::take(&mut buf))).await;
                        }
                        break;
                    }
                    Ok(n) => {
                        if let Ok(s) = std::str::from_utf8(&temp[..n]) {
                            buf.push_str(s);

                            // Try to find sentence boundaries
                            while let Some(idx) = find_sentence_boundary(&buf, max_chars) {
                                if idx > 0 {
                                    let sentence: String = buf.drain(..idx).collect();
                                    if tx.send(Ok(sentence)).await.is_err() {
                                        return;
                                    }
                                } else {
                                    break;
                                }
                            }
                        }
                    }
                    Err(e) => {
                        let _ = tx.send(Err(SegmentError::Io(e))).await;
                        break;
                    }
                }
            }
        });

        Ok(Box::new(DefaultSentenceIterator {
            rx: Arc::new(Mutex::new(rx)),
            closed: false,
        }))
    }
}

/// Finds the index of the first sentence boundary in the string.
/// Returns None if no boundary is found.
fn find_sentence_boundary(s: &str, max_chars: usize) -> Option<usize> {
    let chars: Vec<char> = s.chars().collect();

    // If we've exceeded max chars, force a split
    if chars.len() >= max_chars {
        // Try to find a boundary first
        for (i, &c) in chars.iter().enumerate().take(max_chars) {
            if is_sentence_boundary(c, i, &chars) {
                return Some(s.char_indices().nth(i + 1).map(|(idx, _)| idx).unwrap_or(s.len()));
            }
        }
        // No boundary found, force split at max_chars
        return Some(s.char_indices().nth(max_chars).map(|(idx, _)| idx).unwrap_or(s.len()));
    }

    // Look for natural boundaries
    for (i, &c) in chars.iter().enumerate() {
        if is_sentence_boundary(c, i, &chars) {
            return Some(s.char_indices().nth(i + 1).map(|(idx, _)| idx).unwrap_or(s.len()));
        }
    }

    None
}

/// Checks if a character is a sentence boundary.
fn is_sentence_boundary(c: char, idx: usize, chars: &[char]) -> bool {
    let prev = if idx > 0 { chars[idx - 1] } else { ' ' };
    let next = if idx + 1 < chars.len() { chars[idx + 1] } else { ' ' };

    match c {
        // Handle decimal numbers and times (9.9, 10:15)
        '.' | ':' | ',' | '：' => {
            !(next.is_ascii_digit() && prev.is_ascii_digit())
        }
        // Definite sentence boundaries
        '，' | '；' | '。' | '？' | '！' | '…' | '～' | '?' | '!' | '¿' | '¡' | ';' | '~'
        | '\r' | '\n' | '„' | '・' => true,
        _ => false,
    }
}

struct DefaultSentenceIterator {
    rx: Arc<Mutex<mpsc::Receiver<Result<String, SegmentError>>>>,
    closed: bool,
}

#[async_trait::async_trait]
impl SentenceIterator for DefaultSentenceIterator {
    async fn next(&mut self) -> Result<String, SegmentError> {
        if self.closed {
            return Err(SegmentError::Closed);
        }

        let mut rx = self.rx.lock().await;
        match rx.recv().await {
            Some(result) => result,
            None => Err(SegmentError::Done),
        }
    }

    fn close(&mut self) {
        self.closed = true;
    }
}

#[cfg(test)]
mod segment_tests {
    use super::*;

    #[test]
    fn test_is_sentence_boundary() {
        let chars: Vec<char> = "Hello. World".chars().collect();
        assert!(is_sentence_boundary('.', 5, &chars));

        let chars: Vec<char> = "3.14".chars().collect();
        assert!(!is_sentence_boundary('.', 1, &chars)); // Decimal number

        let chars: Vec<char> = "10:30".chars().collect();
        assert!(!is_sentence_boundary(':', 2, &chars)); // Time
    }

    #[test]
    fn test_find_sentence_boundary() {
        assert_eq!(find_sentence_boundary("Hello. World", 256), Some(6)); // After the period
        assert_eq!(find_sentence_boundary("你好。世界", 256), Some(9)); // 3 bytes per char, after 。
        assert_eq!(find_sentence_boundary("No boundary here", 256), None);
    }

    #[tokio::test]
    async fn test_default_segmenter() {
        let segmenter = DefaultSentenceSegmenter::default();
        let text = "Hello. World!";

        // Use tokio's cursor directly (it implements AsyncRead)
        let async_reader = std::io::Cursor::new(text.as_bytes().to_vec());
        let mut iter = segmenter.segment(Box::new(async_reader)).await.unwrap();

        let s1 = iter.next().await.unwrap();
        assert!(s1.contains("Hello"));
    }
}
